Scripts/Runtime/Collections/UnsafeBuffer.cs

```diff
@@ -0,0 +1,550 @@
+        // READERS
+        //*************************************************************************************************************
+        
+        /// <summary>
+        /// A lightweight struct to allow for reader from the collection
+        /// </summary>
+        [BurstCompatible]
+        public struct Reader
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            [NativeDisableUnsafePtrRestriction] private readonly Lane* m_Lanes;
+            
+            /// <summary>
+            /// How many lanes the collection has available to read from
+            /// </summary>
+            public int LaneCount
+            {
+                get => m_BufferInfo->LaneCount;
+            }
+            
+            internal Reader(ref UnsafeBuffer<T> unsafeBuffer)
+            {
+                m_Lanes = unsafeBuffer.m_Lanes;
+                m_BufferInfo = unsafeBuffer.m_BufferInfo;
+            }
+            
+            /// <summary>
+            /// Creates a <see cref="LaneReader"/> instance locked to a specific lane index.
+            /// </summary>
+            /// <param name="laneIndex">The lane index to use</param>
+            /// <returns>A <see cref="LaneReader"/> instance</returns>
+            public LaneReader CreateReaderForLane(int laneIndex)
+            {
+                Assert.IsTrue(laneIndex < m_BufferInfo->LaneCount && laneIndex >= 0);
+                
+                Lane* lane = m_Lanes + laneIndex;
+                return new LaneReader(lane, m_BufferInfo);
+            }
+        }
+        
+        /// <summary>
+        /// A lightweight reader instance locked to a specific lane.
+        /// <see cref="LaneReader"/>'s manage their own state so that multiple readers can read from a given lane at
+        /// the same time.
+        /// </summary>
+        [BurstCompatible]
+        public struct LaneReader
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            [NativeDisableUnsafePtrRestriction] private readonly Lane* m_Lane;
+            
+            private BlockInfo* m_CurrentReadBlock;
+            private byte* m_EndOfCurrentBlock;
+            private byte* m_ReaderHead;
+            
+            /// <summary>
+            /// How many elements are available to read on this lane
+            /// </summary>
+            public int Count
+            {
+                get => m_Lane->Count;
+            }
+            
+            internal LaneReader(Lane* lane, BufferInfo* bufferInfo)
+            {
+                m_Lane = lane;
+                m_BufferInfo = bufferInfo;
+                
+                m_CurrentReadBlock = m_Lane->FirstBlock;
+                m_ReaderHead = (m_CurrentReadBlock != null) 
+                    ? m_CurrentReadBlock->Data
+                    : null;
+                m_EndOfCurrentBlock = m_ReaderHead + m_BufferInfo->BlockSize;
```

Comment:
How does this work when `m_ReaderHead == null` does that just resolve to 0 or some random memory address?

Reply:
Yes it becomes `0`